### PR TITLE
Rollover

### PR DESCRIPTION
**Pull request by edengal**

### CODE REVIEW

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 1]
```diff
@@ -198,6 +234,37 @@ void index(Map<String, Task> tasksMap) {
         }
     }
 
+    void rolloverIndex(String timbermillAlias) {
+        RolloverRequest rolloverRequest = new RolloverRequest(timbermillAlias, null);
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: if it's the same request every time we can generate it once and store as a final member

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 2]
```diff
@@ -198,6 +234,37 @@ void index(Map<String, Task> tasksMap) {
...
+
+                new Thread(() -> handlePartialTasks(timbermillAlias)).start();
+            }
+        } catch (IOException e) {
+            LOG.error("An error occurred while rollovered index " + timbermillAlias, e);
+        }
+    }
+
+        private void handlePartialTasks(String index){
+            try {
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: add prints for start/end of processing so we know when the thread is done

File: timbermill-client/src/main/java/com/datorama/timbermill/common/TimbermillUtils.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+public class TimbermillUtils {
+	public static final String CLIENT = "client";
+	public static final int THREAD_SLEEP = 2000;
+	private static final Logger LOG = LoggerFactory.getLogger(TimbermillUtils.class);
+	private static String mergingCronExp;
+	private static Set<String> envsSet = Sets.newConcurrentHashSet();
+
+	public static TaskIndexer bootstrap(ElasticsearchParams elasticsearchParams, ElasticsearchClient es) {
+		mergingCronExp = elasticsearchParams.getMergingCronExp();
+		es.bootstrapElasticsearch(elasticsearchParams.getNumberOfShards(), elasticsearchParams.getNumberOfReplicas());
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/common/TimbermillUtils.java]: let's comment it out until we get it working

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 4]
```diff
@@ -206,41 +273,217 @@ void index(Map<String, Task> tasksMap) {
...
 
-    private String getTaskIndexWithEnv(String indexPrefix, String env, ZonedDateTime startTime) {
-        DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern("dd-MM-yyyy");
-        return indexPrefix + INDEX_DELIMITER + env + INDEX_DELIMITER + startTime.format(dateTimeFormatter);
+    public Map<String, Task> getIndexPartialTasks(String index) throws IOException {
+        Scroll scroll = new Scroll(TimeValue.timeValueMinutes(1L));
+        SearchRequest searchRequest = new SearchRequest(index);
+        searchRequest.scroll(scroll);
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        QueryBuilder query = new TermsQueryBuilder("status.keyword", Task.TaskStatus.PARTIAL_ERROR, Task.TaskStatus.PARTIAL_INFO_ONLY, Task.TaskStatus.PARTIAL_SUCCESS, Task.TaskStatus.UNTERMINATED); //todo change keyword
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: status.keyword to const

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 5]
```diff
@@ -206,41 +273,217 @@ void index(Map<String, Task> tasksMap) {
...
+        searchSourceBuilder.query(query);
+        searchSourceBuilder.size(10000);
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: why 10,000? what happens if there are more?

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 6]
```diff
@@ -198,6 +234,37 @@ void index(Map<String, Task> tasksMap) {
...
+        } catch (IOException e) {
+            LOG.error("An error occurred while rollovered index " + timbermillAlias, e);
+        }
+    }
+
+        private void handlePartialTasks(String index){
+            try {
+                Map<String, Task> previousIndexPartialTasks = getIndexPartialTasks(oldIndex);
+                if (!previousIndexPartialTasks.isEmpty()) {
+                    index(previousIndexPartialTasks, index);
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: need to delete from old index

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java [hunk 7]
```diff
@@ -206,41 +273,217 @@ void index(Map<String, Task> tasksMap) {
...
+    public void deleteExpiredTasks() {
+        LOG.info("About to delete expired tasks");
+        RangeQueryBuilder rangeQueryBuilder = new RangeQueryBuilder(ElasticsearchClient.TTL_FIELD).lte(new DateTime());
+        deleteByQuery("_all", rangeQueryBuilder);
     }
+
+    private void deleteByQuery(String index, QueryBuilder query) {
+        DeleteByQueryRequest request = new DeleteByQueryRequest(index).setQuery(query);
+        RequestOptions.Builder builder = RequestOptions.DEFAULT.toBuilder();
+        builder.addHeader("wait_for_completion", "false");
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: wait_for_completion to const

File: timbermill-client/src/main/java/com/datorama/timbermill/TimberLogger.java [hunk 8]
```diff
@@ -58,7 +60,22 @@ public static String start(String name, String parentTaskId, LogParams logParams
 		if (logParams == null){
 			logParams = LogParams.create();
 		}
-		return EventLogger.get().startEvent(name, parentTaskId, logParams);
+		return EventLogger.get().startEvent(name, parentTaskId, logParams, null);
+	}
+
+	public static String startWithDateToDelete(String name, ZonedDateTime dateToDelete) {
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/TimberLogger.java]: startWithDaysRotation

File: timbermill-client/src/main/java/com/datorama/timbermill/cron/TasksMergerJobs.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+		else{
+			LOG.error("Old index doesn't exists, will not merge partial tasks");
+		}
+	}
+
+	private boolean indexExists(String index) {
+		return index != null;
+	}
+
+	private int migrateOldPartialTaskToNewIndex(String currentIndex, String previousIndex) throws IOException {
```
Comment (DvirKatz) [inline timbermill-client/src/main/java/com/datorama/timbermill/cron/TasksMergerJobs.java]: check if we can use the same code for this and for when we create a new index

File: timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java (hunk 1 above)
Comment (edengal) [reply timbermill-client/src/main/java/com/datorama/timbermill/ElasticsearchClient.java]: It' changes per env

**Pull request merged**